crypto/bls/bls.go

```diff
@@ -18,14 +26,33 @@ func RandPrivateKey() *bls.SecretKey {
 	return &sec
 }
 
+var (
+	errEmptyInput = errors.New("BytesToBLSPublicKey: empty input")
+	errPubKeyCast = errors.New("BytesToBLSPublicKey: cast error")
+)
+
 // BytesToBLSPublicKey converts bytes into bls.PublicKey pointer.
 func BytesToBLSPublicKey(bytes []byte) (*bls.PublicKey, error) {
 	if len(bytes) == 0 {
-		return nil, fmt.Errorf("[BytesToBLSPublicKey] bytes is empty")
+		return nil, errEmptyInput
+	}
+	kkey := string(bytes)
+	if k, ok := BLSPubKeyCache.Get(kkey); ok {
+		if pk, ok := k.(bls.PublicKey); ok {
+			return &pk, nil
+		} else {
+			return nil, errPubKeyCast
+		}
 	}
 	pubKey := &bls.PublicKey{}
 	err := pubKey.Deserialize(bytes)
-	return pubKey, err
+
+	if err != nil {
```

Comment:
This is also open to attack. Can we also cache the error?

Reply:
Ummm, this seems cannot be fully mitigated until we implement the blacklist feature. So seems good now.